### ISSUE TITLE
error: expected body of lambda expression

### ISSUE THREAD

**Issue by ryandesign** [ccache/ccache#695]
Hi, after applying the fix for #689 I see a different build failure on OS X 10.8 or 10.9.

### How to reproduce ###

1. build ccache 4.0 + the fix for #689 on OS X 10.8 with `CXXFLAGS=-stdlib=libc++` or 10.9.

### Actual behavior ###

```
In file included from unittest/main.cpp:26:
unittest/../src/third_party/doctest.h:6162:72: error: expected body of lambda expression
    auto DOCTEST_FIX_FOR_MACOS_LIBCPP_IOSFWD_STRING_LINK_ERRORS = []() DOCTEST_NOINLINE
                                                                       ^
unittest/../src/third_party/doctest.h:305:26: note: expanded from macro 'DOCTEST_NOINLINE'
#define DOCTEST_NOINLINE __attribute__((noinline))
                         ^
1 error generated.
```

### Expected behavior ###

successful compile

### Environment ###

ccache 4.0

OS X Mavericks v10.9.5 (13F1911)
Xcode v6.2 (6C131e)
Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)

OS X Mountain Lion v10.8.5 (12F2560)
Xcode v5.1.1 (5B1008)
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)

**Comment by AlexanderLanin** [comment]
Same happens on clang 3.4, because doctest does not support clang 3.4.
I did not look into the actual doctest problem yet.

However I'm surprised you have issues with clang 3.5 (Apple LLVM 6.0).
Did you compile with `-DWARNINGS_AS_ERRORS=OFF` ?

**Comment by AlexanderLanin** [comment]
One more question: Those OS X Versions are way way past their end of support dates, aren't they? Why do you care about them?
https://en.wikipedia.org/wiki/MacOS_version_history#Releases

**Comment by ryandesign** [comment]
> However I'm surprised you have issues with clang 3.5 (Apple LLVM 6.0).

Note that Apple LLVM 6.0 is not equivalent to clang 3.5. It is "based on LLVM 3.5svn", meaning based on some unknown point after the release of clang 3.4 but before the release of 3.5, with an unknown number of further additions and/or subtractions made by Apple.

> Did you compile with `-DWARNINGS_AS_ERRORS=OFF` ?

I wasn't aware of that option. I tried it now and it didn't make a difference. Looks like `OFF` is its default value.

> Those OS X Versions are way way past their end of support dates, aren't they? Why do you care about them?

Yes, but what Apple does or does not support is not relevant to whether some third-party software can or cannot compile or run on it. MacPorts still runs on Mac OS X 10.4 Tiger and later, and we do still have some users using such old versions of macOS (perhaps because they still have an old computer that works fine but cannot be upgraded to a newer OS version). Many of our ports cannot compile on a system so old but others can and if I notice a build failure I try to spend a moment to see if it can be fixed. It may be the case that users of older computers where compiling a program takes hours rather than minutes have the most to gain from ccache.

I already know that ccache 4.0 compiles fine on systems as old as Mac OS X 10.6 if I use a newer open source clang compiler, so I can use that workaround if I need to. But I wanted to do the developers of ccache the courtesy of reporting the problem to them so that they have an opportunity to fix the problem if they wish to do so, rather than keep the knowledge of the problem and potential solutions to myself.

If the developers of ccache don't wish to invest time in supporting older compilers that's fine but in that case it would be helpful if the minimum compiler requirement were documented and/or if the build system detected a too-old compiler and told the user about it rather than failing to build with an obscure error later on. At present, the only compiler requirement mentioned by doc/INSTALL.md is "A C++11 compiler" and "A C99 compiler" which Apple LLVM/clang 5 and above are.

**Comment by AlexanderLanin** [comment]
@jrosdahl doctest did not receive a single commit in the last 4 months. Feels like it will not get us anywhere to request clang 3.4 support there. Maybe we should go back to catch2?

**Comment by jrosdahl** [comment]
@ryandesign: Thanks for your well written bug reports. 

> @jrosdahl doctest did not receive a single commit in the last 4 months. Feels like it will not get us anywhere to request clang 3.4 support there. Maybe we should go back to catch2?
> 
> edit: Catch2 CI does not include clang 3.4 or even 3.5, so it might be even worse there.

(If the choice is between supporting legacy Clang versions and continuing to use Doctest, I would choose the latter.)

> If the developers of ccache don't wish to invest time in supporting older compilers that's fine but in that case it would be helpful if the minimum compiler requirement were documented

It's documentated that Clang 3.8+ is required for building ccache on https://ccache.dev/platform-compiler-language-support.html but I realize that it's not very evident if you start from INSTALL.md. I'll add a reference to that page from INSTALL.md.

I'm of course OK with supporting older versions if fixes are easy and non-intrusive enough .

@ryandesign: How about building the "ccache" target instead of "all" to avoid building the unit tests?

**Comment by AlexanderLanin** [comment]
> (If the choice is between supporting legacy Clang versions and continuing to use Doctest, I would choose the latter.)
I don't really mind, but would still like to understand why?

If it's compile times, during looking deeper into this I saw that catch2 is currently in the process of releasing a version which is supposed to significantly improve compile times (5x).

I personally actually see the bigger problem in not using latest C++ features because we have some ancient compiler in CI. It's always something small and not worth dropping compiler support. But all those small old code fragments end up in old-style code.

**Comment by jrosdahl** [comment]
> > (If the choice is between supporting legacy Clang versions and continuing to use Doctest, I would choose the latter.)
> 
> I don't really mind, but would still like to understand why?

Here's my reasoning:

* The lower bar is Ubuntu 14.04 and RedHat/CentOS 7. This means not using newer features than those implemented in GCC 4.8.4.
* Supporting building with older Clang versions is a nice bonus but not a primary target.
* Catch2 v2 is painfully slow.

> If it's compile times, during looking deeper into this I saw that catch2 is currently in the process of releasing a version which is supposed to significantly improve compile times (5x).

As I understood the [blog post about version 3](https://codingnest.com/the-future-of-catch2/) it will also drop support for GCC 4.8, which makes it unfit for ccache. If version 3 actually will support GCC 4.8.4 (I haven't checked) then using it would be perfectly OK.

> I personally actually see the bigger problem in not using latest C++ features because we have some ancient compiler in CI. It's always something small and not worth dropping compiler support. But all those small old code fragments end up in old-style code.

It sounds like it would be a good idea to try to accept that ccache won't be the place where you can use the latest C++ features. :slightly_smiling_face: The reason why there is an ancient compiler in CI is because I want to continue supporting Linux distributions that some people still use, not because it's fun to support old compilers.

**Comment by AlexanderLanin** [comment]
> * Supporting building with older Clang versions is a nice bonus but not a primary target.

Maybe this will be possible after all with #708
